pkgs/tools/networking/smartdns/default.nix

```diff
@@ -0,0 +1,34 @@
+
+{ stdenv, fetchFromGitHub, openssl }:
+
+stdenv.mkDerivation rec {
+  pname = "smartdns";
+  version = "29";
+
+  src = fetchFromGitHub {
+    owner = "pymumu";
+    repo = pname;
+    rev = "Release${version}";
+    sha256 = "06kgmahgma0ybhacqzj5sfazmhzs3p27a0sxik4xbbk3dyqn9ykj";
+  };
+
+  buildInputs = [ openssl ];
+
+  sourceRoot = "source/src";
+
+  installPhase = "install -Dm755 smartdns $out/bin/smartdns";
```

Comment:
`installPhase` should start with `runHook preInstall` and end with `runHook postInstall`.

Reply:
Could you also install the service file? Will need to be patched similarly to https://git.archlinux.org/svntogit/community.git/tree/trunk/systemd.patch?h=packages/smartdns

Reply:
> Could you also install the service file? Will need to be patched similarly to https://git.archlinux.org/svntogit/community.git/tree/trunk/systemd.patch?h=packages/smartdns

I manually wrote the service inside module.

Reply:
> `installPhase` should start with `runHook preInstall` and end with `runHook postInstall`.

But I don't see the point there, since literally the `installPhase` is quite trivial.

Reply:
It is better to use as much upstream configs as possible to reduce the maintenance effort (the upstream service file will always have the recommended settings).

It would be even better if upstream used a regular build tool like Meson, Autotools or CMake that would install all the files to their appropriate locations and patch the paths in the service file according to the chosen `--prefix` so we did not even have to maintain our own `installPhase`.

Reply:
> It is better to use as much upstream configs as possible to reduce the maintenance effort (the upstream service file will always have the recommended settings).
> 
> It would be even better if upstream used a regular build tool like Meson, Autotools or CMake that would install all the files to their appropriate locations and patch the paths in the service file according to the chosen `--prefix` so we did not even have to maintain our own `installPhase`.

`installPhase` is required since upstream has not taken steps into such tools. I will try to use upstream service file instead but would you like to give me an example package? It will be gladly.

Reply:
> > `installPhase` should start with `runHook preInstall` and end with `runHook postInstall`.
> 
> But I don't see the point there, since literally the `installPhase` is quite trivial.

This is done so that when user does `smartdns.overrideAttrs (attrs: { postInstall = ''do something with the installation''; })` it still works.

Reply:
> > > `installPhase` should start with `runHook preInstall` and end with `runHook postInstall`.
> > 
> > 
> > But I don't see the point there, since literally the `installPhase` is quite trivial.
> 
> This is done so that when user does `smartdns.overrideAttrs (attrs: { postInstall = ''do something with the installation''; })` it still works.

Got it.

Reply:
However I also need to pipe in the customized configuration for smartdns via modification on `script` or `ExecStart`, and packages like `stubby` or `dnsmasq` also do so. Both of them have not used the service file provided by upstream. And I don't see how to do both, additional patch doesn't seem to work.

Reply:
You can just choose to override `ExecStart` in the module. Or use `environment.etc."smartdns/smartdns.conf".source = confFile;` instead of passing the config path as an argument.

Reply:
I have committed it and resolved all the changes